mautic/mautic
# Fixes several bugs with newly introduced device tracking and created source records

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | y
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This addresses several issues found with 2.13.0 beta around tracking contacts with email clickthroughs, dynamic web content, creating audit log entries for new visitors, and recording created source records.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:

A. This PR ensures that email clickthroughs generate tracking cookies based on the email stat than the value of `lead` stored in the clickthrough array and that a tracked identified contact is not merged with a clickthrough contact if different

1. Create a Mautic landing page or set up tracking through a 3rd party page
2. Create an email that links to the landing page
3. Send the email to two contacts
4. Visit the web page in a guest session or a different browser to ensure you are tracked as a visitor
5. view the developer console of the browser to see the cookies you are being tracked with
6. Open the email for one contact and copy the clickthrough URL 
7. Paste it the guest session/different browser and go to it
8. Check the cookies and they are still the visitor ID 
9. Check that the visitor was merged into the contact from the email (the first page hit should be in the sent email contact's timeline)
10. Now copy the clickthrough link from the second contact's email
11. Paste it and go to it in the guest session/separate browser
12. Note the cookies should have been updated to the contact from the second email
13. Check and note that the two identified contacts were merged

B. This PR fixed an issue where first time tracked visitors did not get a created audit log entry and the mtc_device_id and mtc_sid stored in local storage are null

1. Be sure you're working from a clean session in a guest session or different browser than logged into Mautic
2. Visit a page tracked by Mautic (landing page or tracking code)
3. Open the developer console and view the localStorage values 
4. Notice that mtc_id and mtc_device_id are null
5. Note the mtc_id and view the contact in Mautic. Notice that the created entry in the timeline will always be the current date/time because no entry was saved to the database (can verify by viewing the audit log and filtering object_id = the contact ID

C. 3rd party pages with dynamic web content embedded did not have a created by source logged in the lead event table

1. Create a dynamic web content per https://mautic.org/docs/en/dwc/index.html
2. Be sure you're working from a clean session in a guest session or different browser than logged into Mautic
3. Browse to the page with dynamic web content embedded
4. Look for the contact ID in the cookies
5. View the contact's timeline in Mautic
6. Note that there is no "created by source" entry

#### Steps to test this PR:
A. This PR ensures that email clickthroughs generate tracking cookies based on the email stat than the value of `lead` stored in the clickthrough array and that a tracked identified contact is not merged with a clickthrough contact if different

1. Create a Mautic landing page or set up tracking through a 3rd party page
2. Create an email that links to the landing page
3. Send the email to two contacts
4. Visit the web page in a guest session or a different browser to ensure you are tracked as a visitor
5. view the developer console of the browser to see the cookies you are being tracked with
6. Open the email for one contact and copy the clickthrough URL 
7. Paste it the guest session/different browser and go to it
8. Check the cookies and they should be updated to the contact from the email
9. Check that the visitor was merged into the contact from the email (the first page hit should be in the sent email contact's timeline)
10. Now copy the clickthrough link from the second contact's email
11. Paste it and go to it in the guest session/separate browser
12. Note the cookies should have been updated to the contact from the second email
13. Check that the two identified contacts were _NOT_ merged

B. This PR fixed an issue where first time tracked visitors did not get a created audit log entry and the mtc_device_id and mtc_sid stored in local storage are null

1. Be sure you're working from a clean session in a guest session or different browser than logged into Mautic
2. Visit a page tracked by Mautic (landing page or tracking code)
3. Open the developer console and view the localStorage values 
4. Notice that mtc_id and mtc_device_id are populated
5. Note the mtc_id and view the contact in Mautic. Notice that the created entry in the timeline will have a fixed created entry (confirm by viewing the audit lot table in the database)


C. 3rd party pages with dynamic web content embedded did not have a created by source logged in the lead event table

1. Create a dynamic web content per https://mautic.org/docs/en/dwc/index.html
2. Be sure you're working from a clean session in a guest session or different browser than logged into Mautic
3. Browse to the page with dynamic web content embedded
4. Look for the contact ID in the cookies
5. View the contact's timeline in Mautic
6. Note that there is a "created by source" entry recorded as the URL
